### ISSUE TITLE
[Bug]: torchmetrics using `torch.sigmoid()` behind the scenes

### ISSUE THREAD

**Issue by jpcbertoldo** [openvinotoolkit/anomalib#2337]
### Describe the bug

I hope I am wrong on this one :sweat_smile:, but there may be some models with inconsistent F1, F1Max, and AUROC scores\*.

\* Maybe others, but I only checked these.

The documentations of [`torchmetrics.classification.BinaryF1Score`](https://lightning.ai/docs/torchmetrics/stable/classification/f1_score.html#:~:text=shape%20(N%2C%20...).-,If%20preds%20is%20a%20floating%20point%20tensor%20with%20values%20outside%20%5B0%2C1%5D%20range%20we%20consider%20the%20input%20to%20be%20logits%20and%20will%20auto%20apply%20sigmoid%20per%20element.,-Additionally%2C%20we%20convert%20to%20int), [`torchmetrics.classification.BinaryROC`](https://lightning.ai/docs/torchmetrics/stable/classification/roc.html#torchmetrics.classification.BinaryROC), and [`BinaryPrecisionRecallCurve`](https://lightning.ai/docs/torchmetrics/stable/classification/precision_recall_curve.html#torchmetrics.classification.BinaryPrecisionRecallCurve) (used for `F1Max`) say

> ... If preds is a floating point tensor with values outside [0,1] range we consider the input to be logits and will auto apply sigmoid per element. ...

It can be seen [here for instance](https://github.com/Lightning-AI/torchmetrics/blob/1baf920ce8caa24454e9e2e50c075a84f8d7f9ea/src/torchmetrics/functional/classification/stat_scores.py#L110), coming from `BinaryF1Score`.

This affects the metrics because anomaly scores (AS) become "too big", there is a loss of number precision and at some point they are not distinguishable. 

This quick test shows that an AS value 16 would not be ditinguishable of numbers higher than that.

```python
t1 = torch.tensor(16)
t2 = t1 + 1e-1
s1, s2 = t1.sigmoid(), t2.sigmoid()
print(t1 == t2)  # False
print(s1 == s2)  # True
```

This is problematic because many AD models don't necessarily have an apriori range close to zero (where the resolution is unaffected).

For instance, a gaussian model using the mahalanobis distance as AS would output scores proprotional to the square root of the size of the feature space. Embeddings with 300-ish features would be totally undistinguishable because the normal instances would have an average AS `>16`.

I think this might cause a very silent mistake that most users won't see.

> The default engine normalizes the outputs so it's ok in that case, but normalization is not necessarily used.
> Btw why does the normalization happen per batch?
 

### Dataset

N/A

### Model

Other (please specify in the field below)

### Steps to reproduce the behavior

N/A

### OS information

OS information:
- OS: Ubuntu 20.04
- Python version: 3.10.14
- Anomalib version: 1.2.0dev
- PyTorch version: 2.4.0+cu121

### Expected behavior

The metrics in anomalib ideally should find a way to not reach the `.sigmoid()` under the hood of `torchmetrics`.

### Pip/GitHub

pip

### What version/branch did you use?

main

### Configuration YAML

```yaml
n/a
```

### Logs

```shell
n/a
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by alexriedel1** [comment]
For cases wehre the Binary Precision Recall Curve is computed, anomalib uses it`s modified implementation without applying a sigmoid function. So F1Max is safe..

https://github.com/openvinotoolkit/anomalib/blob/2ba5ede95daa2abfdb2026ec08f29f4e8a818f67/src/anomalib/metrics/precision_recall_curve.py#L18-L19

**Comment by jpcbertoldo** [comment]
> For cases wehre the Binary Precision Recall Curve is computed, anomalib uses it`s modified implementation without applying a sigmoid function. So F1Max is safe..
> 
> https://github.com/openvinotoolkit/anomalib/blob/2ba5ede95daa2abfdb2026ec08f29f4e8a818f67/src/anomalib/metrics/precision_recall_curve.py#L18-L19

A ha, missed that. 

It looks like correcting it for ROC would be very similar, right?

anomalib's `BinaryPrecisionRecallCurve` is basically imitating the one from torchmetrics without that `sigmoid()` call in 
`_binary_precision_recall_curve_format`.

**Comment by samet-akcay** [comment]
@jpcbertoldo closing this as Anomalib uses its customized metrics to not apply sigmoid
